core/unit_test/TestGraph.hpp

```diff
@@ -21,6 +21,7 @@
 
 namespace Test {
 
+#if !defined(KOKKOS_ENABLE_SYCL) || defined(SYCL_EXT_ONEAPI_GRAPH)
```

Comment:
Shouldn't the logic be in the `CMake` itself ? Something like:
```cmake
if(KOKKOS_ENABLE_SYCL AND SYCL_EXT_ONEAPI_GRAPH)
    .... enable the SYCL graph test ...
endif()
```

Reply:
The condition reads: only enable the test if the SYCL backend is not enabled, or if the SYCL backend is enabled and the SYCL_EXT_ONEAPI_GRAPH feature macro is defined.

Reply:
> Shouldn't the logic be in the `CMake` itself ?

The feature macro is not a `CMake` variable.

Reply:
OK. I would suggest to move the line
```c++
#if !defined(KOKKOS_ENABLE_SYCL) || defined(SYCL_EXT_ONEAPI_GRAPH)
```
to `TestSYCL_Graph.hpp` then.

Reply:
We typically auto-generate the *.cpp test files (such as `TestSYCL_Graph.cpp`) which we might want to do in a follow-up. Moving the guard to the *.cpp file would prevent that.

Reply:
OK. I would strongly advise a "CMake experimental feature flag" like `KOKKOS_ENABLE_SYCL_GRAPH` or so for several reasons :1234: 

1. I feel strongly against introducing backend-specific preprocessing guard that deactivate the whole bunch of **generic tests** in `TestGraph.hpp`.
   This goes against the idea of a portable test file and is a backdoor for future maintenance headaches.
2. It feels strange that if a user has the `SYCL` backend enabled, but hasn't support for `SYCL_EXT_ONEAPI_GRAPH`, still there will be a test target generated
   for `TestSYCL_Graph.cpp` that actually does nothing. To me, such a test deactivation should be done in the `CMakeLists.txt` for sure.
3. Am I right to say that if `KOKKOS_ENABLE_SYCL` is defined, but `SYCL_EXT_ONEAPI_GRAPH` is not, then `Kokkos::Graph<SYCL>` will fallback on the
   default `GraphImpl`? If so, I see that as a very big **warning** sign :warning:. Indeed, as a user that wants the `SYCL` specialization, I would not be happy to see that
   `Kokkos::Graph<SYCL>` silently falls back on the default `Graph` implementation...

For these reasons, I strongly suggest that you introduce `KOKKOS_ENABLE_SYCL_GRAPH` or so. It would be an "experimental feature" flag that, if set by the user,
enforces the `SYCL` `Graph` specialization (and therefore ensures that compilation fails if `SYCL_EXT_ONEAPI_GRAPH` is not supported).

You could also check for the `SYCL_EXT_ONEAPI_GRAPH` support with some kind of `try_compile`.

The logic could be align these lines:
```cmake
if(NOT DEFINED KOKKOS_ENABLE_SYCL_GRAPH)
    try_compile(...)
    if(COMPILATION_SUCCEEDED)
        set(KOKKOS_ENABLE_SYCL_GRAPH ON)
    endif()
endif()

...

if(KOKKOS_ENABLE_SYCL_GRAPH)
    add_test(TestSYCL_Graph.cpp)
endif()
```

What do you think ?

Reply:
> 1. I feel strongly against introducing backend-specific preprocessing guard that deactivate the whole bunch of generic tests in TestGraph.hpp.
> 2. It feels strange that if a user has the SYCL backend enabled, but hasn't support for SYCL_EXT_ONEAPI_GRAPH, still there will be a test target generated
for TestSYCL_Graph.cpp that actually does nothing. To me, such a test deactivation should be done in the CMakeLists.txt for sure.

We shouldn't actually need that guard and can just always execute the test even if no specialization is available, see #7011.

> Am I right to say that if KOKKOS_ENABLE_SYCL is defined, but SYCL_EXT_ONEAPI_GRAPH is not, then Kokkos::Graph<SYCL> will fallback on the
default GraphImpl? If so, I see that as a very big warning sign ⚠️. Indeed, as a user that wants the SYCL specialization, I would not be happy to see that
Kokkos::Graph<SYCL> silently falls back on the default Graph implementation...

That's how the Graph implementation is designed and we should not start breaking Graphs for users if there is a specialization but it's not active in their configuration. We always reserve the right to improve performance based on the backend capability without the need for the user to pitch in. I added information about `SYCL_EXT_ONEAPI_GRAPH` to `print_configuration`, though.

Reply:
> We shouldn't actually need that guard and can just always execute the test even if no specialization is available, see https://github.com/kokkos/kokkos/pull/7011.

Nice :+1:

> That's how the Graph implementation is designed and we should not start breaking Graphs for users if there is a specialization but it's not active in their configuration. We always reserve the right to improve performance based on the backend capability without the need for the user to pitch in. I added information about SYCL_EXT_ONEAPI_GRAPH to print_configuration, though.

My suggestion is about allowing the user to "fore opt-in into the specialization or stop compilation" :wink: So if I define `KOKKOS_ENABLE_SYCL_GRAPH_IMPL`, and for some reason it will not be possible to compile the `SYCL` specialization, I would prefer that `Kokkos` does not compile, rather than falling back to the default implementation. However, if I don't define `KOKKOS_ENABLE_SYCL_GRAPH_IMPL`, then I agree with you that `Kokkos` can do whatever it want.

Reply:
I don't understand what your use case is. Why would you rather fail compilation than use a fallback implementation? Since `KOKKOS_ENABLE_SYCL_GRAPH_IMPL` is not portable, you can just as well check for `SYCL_EXT_ONEAPI_GRAPH` in user code.

Note that we already have
```
#if !((HIP_VERSION_MAJOR == 5) && (HIP_VERSION_MINOR == 2))
#define KOKKOS_IMPL_HIP_GRAPH_ENABLED
#endif
```
which silently enables or disables the Graph specialization based on the compiler version.